runners/jstorm/src/main/java/com/alibaba/jstorm/beam/translation/runtime/state/JStormBagState.java

```diff
@@ -0,0 +1,177 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.jstorm.beam.translation.runtime.state;
+
+import com.alibaba.jstorm.cache.ComposedKey;
+import com.alibaba.jstorm.cache.IKvStore;
+import org.apache.beam.runners.core.StateNamespace;
+import org.apache.beam.sdk.state.BagState;
+import org.apache.beam.sdk.state.ReadableState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * JStorm implementation of {@link BagState}.
+ */
+class JStormBagState<K, T> implements BagState<T> {
+    private static final Logger LOG = LoggerFactory.getLogger(JStormBagState.class);
+
+    @Nullable
+    private final K key;
+    private final StateNamespace namespace;
+    private final IKvStore<ComposedKey, T> kvState;
+    private final IKvStore<ComposedKey, Object> stateInfoKvState;
+    private int elemIndex;
+
+    public JStormBagState(@Nullable K key, StateNamespace namespace, IKvStore<ComposedKey, T> kvState,
+                           IKvStore<ComposedKey, Object> stateInfoKvState) throws IOException {
+        this.key = key;
+        this.namespace = checkNotNull(namespace, "namespace");
+        this.kvState = checkNotNull(kvState, "kvState");
+        this.stateInfoKvState = checkNotNull(stateInfoKvState, "stateInfoKvState");
+
+        Integer index = (Integer) stateInfoKvState.get(getComposedKey());
```

Comment:
This should be lazy or executed like a future like `ReadableState<Integer>` so there is no cost until it is read or written.

But actually, the whole point of `BagState` is that you can do "blind writes" where you don't do a read. It is good that you don't read the whole bag, but it would be better if there were a way to not read anything.